### PR TITLE
SCP-[4711..4714] Withdrawals REST API

### PR DESCRIPTION
**Pull request by jhbertra**
Adds the full REST API for withdrawals (`GET` (all), `GET` (one), `POST`, `PUT`)

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
        - Review required
        - [x] Substantial changes to code, test, or documentation
    - [x] Reviewer requested

### CODE REVIEW

**Review by ladamesny — APPROVED** [review]
Looks good!

**Comment by ladamesny** [comment]
Sorry I just noticed the ci failed, We'll hold until it's passing again.

File: marlowe-runtime-web/src/Language/Marlowe/Runtime/Web/Types.hs [hunk 1]
```diff
@@ -249,6 +253,42 @@ instance ToJSON ContractHeader
...
+  , roleTokenMintingPolicyId :: PolicyId
+  , role :: Text
+  } deriving (Show, Eq, Ord, Generic)
+
+instance ToJSON PayoutRef
+instance FromJSON PayoutRef
+instance ToSchema PayoutRef
+
+data Withdrawal = Withdrawal
+  { payouts :: Set PayoutRef
```
[the remark this answers is not in the document]
Comment (jhbertra) [reply marlowe-runtime-web/src/Language/Marlowe/Runtime/Web/Types.hs:282]: A withdrawal withdraws one or more role validator payouts. Each payout is a tx output to the payout validator with a role token datum (consisting of the role name and the role token policy ID / currency symbol).

To spend a payout (i.e. withdraw it) in a transaction, that transaction must spend a matching role token in one of its inputs. A single withdraw transaction can spend multiple payouts provided that the required role tokens are present in the inputs. Those payouts can be from different payout validator addresses and for different role tokens.

File: marlowe-runtime-web/server/Language/Marlowe/Runtime/Web/Server/TxClient.hs [hunk 2]
```diff
@@ -128,6 +143,7 @@ txClient :: forall r. Component IO (TxClientDependencies r) (TxClient r)
 txClient = component \TxClientDependencies{..} -> do
   tempContracts <- newTVar mempty
   tempTransactions <- newTVar mempty
+  tempWithdrawals <- newTVar mempty
```
Comment (bwbush) [inline marlowe-runtime-web/server/Language/Marlowe/Runtime/Web/Server/TxClient.hs:146]: So contracts, transactions, and withdrawals are the ephermeral local state of the webserver? Is there some sort of garbage collection on these, or will the maps get larger and larger as clients choose not to sign and submit some transactions? I suppose that unsigned transactions could be garbage collected when their validity interval passes.

**Review by bwbush — APPROVED** [review]
Looks good, but I'm not very familiar with the web server and haven't manually tested.

File: marlowe-runtime-web/server/Language/Marlowe/Runtime/Web/Server/REST/ApiError.hs [hunk 3]
```diff
@@ -109,6 +110,29 @@ rangeNotSatisfiable msg errorCode = toServerError . ApiError msg (fromMaybe "Ran
...
 rangeNotSatisfiable' msg = rangeNotSatisfiable msg Nothing
 
+instance HasDTO (WithdrawError 'V1) where
+  type DTO (WithdrawError 'V1) = ApiError
+
+instance ToDTO (WithdrawError 'V1) where
+  toDTO = \case
+    WithdrawConstraintError (MintingUtxoNotFound _) -> ApiError "Minting UTxO not found" "MintingUtxoNotFound" Null 500
+    WithdrawConstraintError (RoleTokenNotFound _) -> ApiError "Role token not found" "RoleTokenNotFound" Null 403
+    WithdrawConstraintError ToCardanoError -> ApiError "Internal error" "InternalError" Null 500
```
[the remark this answers is not in the document]
Comment (jhbertra) [reply marlowe-runtime-web/server/Language/Marlowe/Runtime/Web/Server/REST/ApiError.hs:120]: Yes, the errors should be better. Eventually, I want a better-designed error response in general.

**Pull request merged**